pkg/runner/resume.go

```diff
@@ -0,0 +1,270 @@
+			}
+
+			if err := runner.Wait(); err != nil {
+				return errors.Join(ErrCouldNotWaitForFirecracker, err)
+			}
+
+			for _, device := range [][3]string{
+				{runner.stateName, stateCopyName, snapshotLoadConfiguration.ExperimentalMapPrivateStateOutput},
+				{runner.memoryName, memoryCopyName, snapshotLoadConfiguration.ExperimentalMapPrivateMemoryOutput},
+			} {
+				inputFile, err := os.Open(filepath.Join(runner.server.VMPath, device[1]))
+				if err != nil {
+					return errors.Join(snapshotter.ErrCouldNotOpenInputFile, err)
+				}
+				defer inputFile.Close()
+
+				var (
+					outputPath = device[2]
+					addPadding = true
+				)
+				if outputPath == "" {
+					outputPath = filepath.Join(runner.server.VMPath, device[0])
+					addPadding = false
+				}
+
+				if err := os.MkdirAll(filepath.Dir(outputPath), os.ModePerm); err != nil {
+					panic(errors.Join(packager.ErrCouldNotCreateOutputDir, err))
+				}
+
+				outputFile, err := os.OpenFile(outputPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, os.ModePerm)
+				if err != nil {
+					return errors.Join(packager.ErrCouldNotOpenOutputFile, err)
+				}
+				defer outputFile.Close()
+
+				deviceSize, err := io.Copy(outputFile, inputFile)
+				if err != nil {
+					return errors.Join(snapshotter.ErrCouldNotCopyFile, err)
+				}
+
+				// We need to add a padding like the snapshotter if we're writing to a file instead of a block device
+				if addPadding {
+					if paddingLength := utils.GetBlockDevicePadding(deviceSize); paddingLength > 0 {
+						if _, err := outputFile.Write(make([]byte, paddingLength)); err != nil {
+							return errors.Join(snapshotter.ErrCouldNotWritePadding, err)
+						}
+					}
+				}
+			}
+		}
+
+		return nil
+	}
+
+	goroutineManager := manager.NewGoroutineManager(
+		ctx,
+		&errs,
+		manager.GoroutineManagerHooks{
+			OnAfterRecover: func() {
+				if suspendOnPanicWithError {
+					suspendCtx, cancelSuspendCtx := context.WithTimeout(runner.rescueCtx, rescueTimeout)
+					defer cancelSuspendCtx()
+
+					// Connections need to be closed before creating the snapshot
+					if resumedRunner.acceptingAgent != nil {
+						if e := resumedRunner.acceptingAgent.Close(); e != nil {
+							errs = errors.Join(errs, snapshotter.ErrCouldNotCloseAcceptingAgent, e)
+						}
+					}
+					if resumedRunner.agent != nil {
+						resumedRunner.agent.Close()
+					}
+
+					// If a resume failed, flush the snapshot so that we can re-try
+					if err := resumedRunner.createSnapshot(suspendCtx); err != nil {
+						errs = errors.Join(errs, ErrCouldNotCreateRecoverySnapshot, err)
+					}
+				}
+			},
+		},
+	)
+	defer goroutineManager.Wait()
+	defer goroutineManager.StopAllGoroutines()
+	defer goroutineManager.CreateBackgroundPanicCollector()()
+
+	// We intentionally don't call `wg.Add` and `wg.Done` here since we return the process's wait method
+	// We still need to `defer handleGoroutinePanic()()` here however so that we catch any errors during this call
```

Comment:
```suggestion
```

Probably fine to remote since it's outdated?

Reply:
Resolved in 88ddb283155e97e3fdae49179ed5c9e3a1c63b23